### PR TITLE
Allow for "zfs receive" to skip existing snapshots

### PR DESCRIPTION
**Pull request by FransUrbo**
by adding the option "-s" to zfs receive.

Signed-off-by: Turbo Fredriksson <turbo@bayour.com>

Closes: #2666

### CODE REVIEW

File: lib/libzfs/libzfs_sendrecv.c [hunk 1]
```diff
@@ -2968,6 +2969,17 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
 	zc.zc_begin_record = drr_noswap->drr_u.drr_begin;
 	zc.zc_cookie = infd;
 	zc.zc_guid = flags->force;
+
+	if (zfs_dataset_exists(hdl, zc.zc_value, ZFS_TYPE_SNAPSHOT) &&
+	    flags->skip) {
+		(void) printf("skipping snapshot %s - %s already exists\n",
+				drrb->drr_toname, zc.zc_value);
```
Comment (behlendorf) [inline lib/libzfs/libzfs_sendrecv.c]: Normally I'd say never to use `printf()` in a library function like this.  Unfortunately, it's already being done in this function so let's just make sure it's wrapped with an `if (flags->verbose)` conditional like the other cases.

File: man/man8/zfs.8 [hunk 2]
```diff
@@ -3017,6 +3017,33 @@ Do not actually receive the stream. This can be useful in conjunction with the \
...
+.sp .6
+.RS 4n
+Generate a more compact stream by using WRITE_EMBEDDED records for blocks
+which are stored more compactly on disk by the \fBembedded_data\fR pool
+feature.  This flag has no effect if the \fBembedded_data\fR feature is
+disabled.  The receiving system must have the \fBembedded_data\fR feature
+enabled.  If the \fBlz4_compress\fR feature is active on the sending system,
+then the receiving system must have that feature enabled as well. See
+\fBzpool-features\fR(5) for details on ZFS feature flags and the
+\fBembedded_data\fR feature.
```
Comment (behlendorf) [inline man/man8/zfs.8]: This unrelated hunk can be dropped, it was added to master in another commit.

File: man/man8/zfs.8 [hunk 3]
```diff
@@ -3017,6 +3017,33 @@ Do not actually receive the stream. This can be useful in conjunction with the \
...
+
+.sp
+.ne 2
+.mk
+.na
+\fB\fB-s\fR\fR
+.ad
+.sp .6
+.RS 4n
+Skip existing snapshots in a stream.
```
Comment (behlendorf) [inline man/man8/zfs.8]: This could use a little more explanation.

File: lib/libzfs/libzfs_sendrecv.c [hunk 4]
```diff
@@ -2874,13 +2874,14 @@ zfs_receive_one(libzfs_handle_t *hdl, int infd, const char *tosnap,
...
 		 * be an incremental, or the stream specifies a new fs
 		 * (full stream or clone) and they want us to blow it
 		 * away (and have therefore specified -F and removed any
 		 * snapshots).
 		 */
-		if (stream_wantsnewfs) {
+		if (stream_wantsnewfs && !flags->skip) {
 			if (!flags->force) {
 				zcmd_free_nvlists(&zc);
 				zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
```
Comment (behlendorf) [inline lib/libzfs/libzfs_sendrecv.c]: Let's update this error message to reference the optional `-s' option.  Something like "destination '%s' exists\nmust specify -F to overwrite it or -s to skip it"

**Comment by kpande** [comment]
I could see a use case for this when one doesn't have a resumable send stream, and would still like to skip existing snaps.

**Comment by rlaager** [comment]
@kpande, can you expand on the use case you are thinking of?

**Comment by behlendorf** [comment]
Closing this issue until someone has the time to work on it.

**Pull request closed without merging**